Move out globalization code from runtime

mkhamoyan opened the pull request:
Currently we compile the globalization code twice: 
1) directly into the mono runtime, by duplicating the logic
2) from src/native/System.Globalization.Native

By this PR we will move the globalization code out of the mono runtime.

mkhamoyan committed:
Move out globalization code from runtime

mkhamoyan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

vargaz commented:
If the code in native-library.c is removed, how are these pinvoke methods going to be resolved ?

mkhamoyan commented:
> If the code in native-library.c is removed, how are these pinvoke methods going to be resolved ?

https://github.com/dotnet/runtime/blob/bd5f6eb73f6dffe91386f75d3ed5531ddcb2aeb6/src/coreclr/vm/pinvokeoverride.cpp#L32 I think this part will handle it

jkoritzinsky commented:
> > If the code in native-library.c is removed, how are these pinvoke methods going to be resolved ?
> 
> https://github.com/dotnet/runtime/blob/bd5f6eb73f6dffe91386f75d3ed5531ddcb2aeb6/src/coreclr/vm/pinvokeoverride.cpp#L32
> 
> I think this part will handle it

This is in the CoreCLR code-base, which isn't used by Mono at all. The code you removed in native-library.c is the equivalent code in Mono.

mkhamoyan committed:
Merge branch 'main' into globalization_move_out

mkhamoyan commented:
> This is in the CoreCLR code-base, which isn't used by Mono at all. The code you removed in native-library.c is the equivalent code in Mono.

I see, but it is not clear for me why after removing code from `native-library.c`, tests which pinvoke globalization code didn't complain.

mkhamoyan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

mkhamoyan commented:
@lewing @akoeplinger could you please review

akoeplinger committed:
Remove last traces of icu code from the runtime

mkhamoyan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

akoeplinger committed:
Remove a few more icu-related logic

mkhamoyan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

mkhamoyan committed:
revert back _MonoCFlags for wasm

mkhamoyan commented on src/mono/mono.proj:
```diff
@@ -444,9 +433,6 @@
       <_MonoCMakeArgs Include="-DEMSCRIPTEN_SYSTEM_PROCESSOR=wasm"/>
-      <_MonoCFLAGS Condition="'$(WasmEnableThreads)' == 'true'" Include="$(EscapedQuoteW)-I$([MSBuild]::NormalizePath('$(PkgMicrosoft_NETCore_Runtime_ICU_Transport)', 'runtimes', 'browser-wasm-threads', 'native', 'include'))$(EscapedQuoteW)"/>
-
-      <_MonoCFLAGS Condition="'$(WasmEnableThreads)' != 'true'" Include="$(EscapedQuoteW)-I$([MSBuild]::NormalizePath('$(PkgMicrosoft_NETCore_Runtime_ICU_Transport)', 'runtimes', 'browser-wasm', 'native', 'include'))$(EscapedQuoteW)"/>
```
There are CI failures, reverted this part to validate if this caused.

mkhamoyan committed:
remove icu related code

mkhamoyan committed:
Merge branch 'main' into globalization_move_out

mkhamoyan committed:
Fix windows failure

mkhamoyan committed:
Fix windows build

mkhamoyan committed:
bring back STATIC_LIBS_ONLY check

mkhamoyan committed:
test

akoeplinger committed:
Merge branch 'main' into globalization_move_out

akoeplinger committed:
Make System.Globalization.Native compile as C on recent MSVC

We can now use __typeof__ in recent MSVC and some small tweaks to how
the function pointers are defined.

Also fix the PDB not getting copied next to the .dll and disable
validating the .so file which doesn't apply to Windows.

akoeplinger committed:
Fix compiler warning

am11 commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -135,8 +131,9 @@ if (NOT CLR_CMAKE_TARGET_BROWSER AND NOT CLR_CMAKE_TARGET_WASI AND NOT CLR_CMAKE
 
-if (MSVC)
-    set_source_files_properties(${NATIVEGLOBALIZATION_SOURCES} PROPERTIES LANGUAGE CXX)
+if (MSVC AND MSVC_VERSION LESS 1939)
```
Our minimal VS version [offsets](https://learn.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=msvc-160#c-standard-library-features-1) C99 support required here, so this check is unnecessary.

https://github.com/dotnet/runtime/blob/79a1c917798a1fa27455a8683657bc2789eee939/eng/native/init-vs-env.cmd#L57

akoeplinger replied:
`__typeof__` support is new in MSVC 19.39: https://learn.microsoft.com/en-us/cpp/c-language/typeof-c?view=msvc-170

mkhamoyan committed:
fix .dll version and add in runtime pack

mkhamoyan committed:
revert type changes

mkhamoyan committed:
Add System.Globalization.Native.def

mkhamoyan committed:
Exclude globalization.native libs when not mono

mkhamoyan committed:
Exclude System.Globalization.Native.dll when not mono

mkhamoyan committed:
Merge branch 'main' into globalization_move_out

mkhamoyan committed:
revert MSVC_VERSION checks

mkhamoyan commented:
/azp run runtime-extra-platforms

mkhamoyan committed:
Merge branch 'main' into globalization_move_out

mkhamoyan commented:
/azp run runtime-extra-platforms

mkhamoyan committed:
Merge branch 'main' into globalization_move_out

mkhamoyan commented:
/azp run runtime-extra-platforms

mkhamoyan commented:
Failures are not related.
